### PR TITLE
[PLT-344] Add dataset.upsert_data_rows method

### PR DESCRIPTION
**Pull request by attila-papai**
This PR adds support for upserting data rows to a dataset, meaning the provided data row specifications either create new data rows or update existing ones. 

Please see the added unit tests for example usages of this new functionality.

### CODE REVIEW

**Review by vbrodsky — COMMENTED** [review]
I have a question regarding Task/returns.

A few months ago, when we were discussing upserts, we talked about returning results, specifically successes and failures (and warnings? partial successes?). I think there was an issue with how create_data_rows currently works, but I don't remember all the details. Oh yeah, one of them was like how many data rows we returns errors? I.e. is there a limit etc Perhaps Matt C. might recall better. Is this issue within the scope of the current implementation to review task return to make it support 'infinite' number of rows?

File: labelbox/schema/dataset.py
Comment (mrobers1982) [inline labelbox/schema/dataset.py:759]: For now, let's also enforce this limit - https://github.com/Labelbox/labelbox-python/blob/develop/labelbox/schema/dataset.py#L34.

This matches the expectation for imports, in the future I can update the backend to remove this limitation.

File: labelbox/schema/dataset.py
Comment (mrobers1982) [inline labelbox/schema/dataset.py:766]: If we re-use this method we need to expose setting the file name; otherwise I believe it will just overwrite the file if there are multiple chunks.

File: labelbox/schema/dataset.py
Comment (mrobers1982) [inline labelbox/schema/dataset.py:783]: We should validate the dataset id set on each specification (if any) matches what is set here.

File: labelbox/schema/data_row.py
Comment (mrobers1982) [inline labelbox/schema/data_row.py:40]: This will ultimately need to support metadata, attachments, embeddings, .etc, essentially all the fields supported by import.

File: labelbox/schema/data_row.py
Comment (mrobers1982) [inline labelbox/schema/data_row.py:47]: We might consider abstracting this concept from the user and map the user-facing entity to this type which the backend expects.

One example is the user can optionally provide a key; otherwise we assume it will be auto-generated (on create).

File: labelbox/schema/dataset.py
Comment (attila-papai) [reply labelbox/schema/dataset.py]: it won't overwrite it, an UUID is appended to the filename as a prefix (in api)

File: labelbox/schema/dataset.py
Comment (attila-papai) [reply labelbox/schema/dataset.py]: We make chunks of these items and upload them to LB as separate JSON files. Then we create a manifest file referencing all the chunk files and also upload the manifest file to LB. Then we call backend with the manifest URL only. The Data Ingestion system will then parallelize the import as much as possible.

File: labelbox/schema/dataset.py
Comment (mrobers1982) [inline labelbox/schema/dataset.py]: Perhaps we should mangle this value by prefixing with `__` - in general we should do better in this project when defining "private" variables.

File: labelbox/schema/data_row.py
Comment (mrobers1982) [inline labelbox/schema/data_row.py]: I'm not sure we should include these models here - they are only used for `upsert` and have very generic names which might lead to confusion on when to use and why and IMO we do not want to set any new standards as part of this change.

We can think on this more but we most likely want to get buy-in from code-owners and/or MLSE.

File: labelbox/schema/dataset.py
Comment (mrobers1982) [inline labelbox/schema/dataset.py:308]: I understand you wanted to keep it [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), but the code is arguably more complex as a result.

File: labelbox/schema/dataset.py
Comment (mrobers1982) [inline labelbox/schema/dataset.py]: I understand functions within functions is a pattern already used in this file but I would avoid this pattern unless there's not viable alternative - there's a performance penalty, not to mention it makes the code more difficult to read.

File: labelbox/schema/dataset.py
Comment (mrobers1982) [inline labelbox/schema/dataset.py]: I understand this is an optimization but did you consider uploading the chunks in parallel?

File: labelbox/schema/data_row.py
Comment (attila-papai) [reply labelbox/schema/data_row.py]: in the follow up refactor ticket, I was gonna reuse this for the create methods
IMO we better start using types to make this SDK library easier to use. Without it, clients need to either read the docs or the underlying code itself, especially when we use `*args` and `**kwargs` that makes it even more complicated to understand what variables are supported.

File: labelbox/schema/dataset.py
Comment (attila-papai) [reply labelbox/schema/dataset.py:308]: I will address this in the follow-up work.

File: labelbox/schema/dataset.py
Comment (attila-papai) [reply labelbox/schema/dataset.py]: I didn't want to use a class in the first place. Sync up with @vbrodsky 
I'm gonna remove it for now.

File: labelbox/schema/data_row.py
Comment (mrobers1982) [reply labelbox/schema/data_row.py]: Can you provide more details re: what you plan to refactor?

I'm worried the scope is too large - you mention refactoring `create_data_rows` but you also need to be mindful that we do not break existing implementations as that would seemingly imply a major upgrade to the SDK.

Personally, I believe we should aim our sights lower with a minimal integration, minimizing changes to existing patterns.

Per above, if you want to continue down this path we likely need to get buy-in from the SDK team as they would likley own this work.

File: labelbox/schema/dataset.py
Comment (mrobers1982) [reply labelbox/schema/dataset.py]: My comment was not so much in the class as the functions within functions, personally I would extract those to avoid the performance penalty as I do not think you need the closure.

File: labelbox/schema/asset_attachment.py
Comment (vbrodsky) [inline labelbox/schema/asset_attachment.py:64]: why not make a method on AttachmentType `valid_attachment` ?

File: labelbox/schema/dataset.py
Comment (vbrodsky) [inline labelbox/schema/dataset.py:57]: how was this number chosen?

File: labelbox/schema/data_row.py
Comment (sfendell-labelbox) [reply labelbox/schema/data_row.py]: I'd love to have types in all these methods, but I think Matt's right; it's not trivial to add typing to these methods without breaking backward compatibility, and until we do that, we'd have 2 similar methods taking in arguments in different formats.

Since we can't add typing to other methods immediately, we should aim for consistency for now. If we do upgrade to typed arguments, we can do it in all the relevant methods at once, in a single PR, and announce it in patch notes for the relevant SDK version.

Could you file a ticket (if we don't have one already) for adding types here and put it in backlog so this doesn't get lost?

File: labelbox/schema/dataset.py
Comment (sfendell-labelbox) [reply labelbox/schema/dataset.py]: I agree solely on readability concerns, no need to nest the function instead of making it a private func if it doesn't use anything from the outer scope.

Or even just not having a function for this at alll; the function is defined inline right before it's used for the only time so it's not like it's encapsulating anything or reducing line count.

File: labelbox/schema/dataset.py
Comment (mrobers1982) [reply labelbox/schema/dataset.py:57]: nit: In Python, constants are typically written in all capital letters.

File: labelbox/schema/dataset.py
Comment (mrobers1982) [inline labelbox/schema/dataset.py]: While it does not look like the code supports this case, I would not give an end-user multiple ways of doing the same thing, it will only create confusion.

File: labelbox/schema/dataset.py
Comment (mrobers1982) [inline labelbox/schema/dataset.py]: There was an ask to remove the nested functions, can you take care of that too?

File: labelbox/schema/dataset.py
Comment (mrobers1982) [inline labelbox/schema/dataset.py]: Personally, I would expose this value in the method signature and provide a reasonable default.

File: labelbox/schema/dataset.py
Comment (mrobers1982) [inline labelbox/schema/dataset.py]: nit: When setting the value in the line above, you can also use `pop` to retrieve the value and delete the key from the dictionary.

**Review by mrobers1982 — CHANGES_REQUESTED** [review]
A couple minor things, but almost there - thanks for taking care of this!

**Review by mrobers1982 — APPROVED** [review]
Let's ultimately wait to merge until we have an approval from @sfendell-labelbox or @vbrodsky.

**Comment by mrobers1982** [comment]
Let's potentially hold off on merging this change if product wants us to move the implementation from `Dataset` to `Client`.

https://labelbox.atlassian.net/browse/PLT-111

**Comment by mrobers1982** [comment]
> Let's potentially hold off on merging this change if product wants us to move the implementation from `Dataset` to `Client`.
> 
> https://labelbox.atlassian.net/browse/PLT-111

~~Actually, let's assume we are moving to the `Client` and make these changes.~~

Product wants to get feedback from stakeholders before we make this change.

**Comment by mrobers1982** [comment]
> > Let's potentially hold off on merging this change if product wants us to move the implementation from `Dataset` to `Client`.
> > https://labelbox.atlassian.net/browse/PLT-111
> 
> ~Actually, let's assume we are moving to the `Client` and make these changes.~
> 
> Product wants to get feedback from stakeholders before we make this change.

Product wants to leave as-is for now - https://labelbox.atlassian.net/browse/PLT-111?focusedCommentId=192239.

We can go ahead and merge if everyone is happy with the current implementation.

**Review by vbrodsky — APPROVED** [review]

**Pull request merged**
